File: Graphs/P019.py
def brute(startWord,targetWord,wordList):
    wordSet=set(wordList)
    minStep=[float('inf')]
    minList=[]
    def diff_by_oneword(w1,w2):
        diff=0
        for i in range(len(w1)):
            if w1[i]!=w2[i]:
                diff+=1
            if diff>1:
                return False
        return diff==1
    def dfs(word,step,visited,path):
        if word==targetWord:
            if step<minStep[0]:
                minStep[0]=step
                minList.clear()
                minList.append(path[:])
            elif step==minStep[0]:
                minList.append(path[:])
            return
        for next in wordList:
            if diff_by_oneword(next,word) and next not in visited:
                visited.add(next)
                path.append(next)
                dfs(next,step+1,visited,path)
                visited.remove(next)
                path.pop()
    dfs(startWord,1,set([startWord]),[startWord])
    return minList

from collections import deque, defaultdict

def optimal(startWord, targetWord, wordList):
    wordSet = set(wordList)
    if targetWord not in wordSet:
        return []
    q = deque([startWord])
    parents = defaultdict(list)
    visited = set([startWord])
    found = False
    while q and not found:
        next_level = set()
        for _ in range(len(q)):
            word = q.popleft()
            for i in range(len(word)):
                for ch in "abcdefghijklmnopqrstuvwxyz":
                    newWord = word[:i] + ch + word[i+1:]
                    if newWord in wordSet and newWord not in visited:
                        parents[newWord].append(word)
                        if newWord == targetWord:
                            found = True
                        next_level.add(newWord)
        q.extend(next_level)
        visited |= next_level
    res = []
    def backtrack(word, path):
        if word == startWord:
            res.append(path[::-1])
            return
        for p in parents[word]:
            backtrack(p, path + [p])
    if found:
        backtrack(targetWord, [targetWord])
    return res

startWord = "hit"
targetWord = "cog"
wordList = ["hot", "dot", "dog", "lot", "log", "cog"]
print("Brute: \n") 
res1 = brute(startWord, targetWord, wordList)
for row in res1:
    print('->'.join(row))
print("\nOptimal: \n") 
res2 = optimal(startWord, targetWord, wordList)
for row in res2:
    print('->'.join(row))

File: Striver_C++_Problems/Graphs/P019.cpp
// P019.cpp - Word Ladder II (all shortest transformation sequences)

#include <bits/stdc++.h>
using namespace std;

bool diff_by_one_2(const string& a, const string& b) {
    int diff = 0;
    for (int i = 0; i < (int)a.size(); i++) {
        if (a[i] != b[i]) {
            diff++;
            if (diff > 1) return false;
        }
    }
    return diff == 1;
}

// Brute: DFS exploring all paths, tracking global minimum length
vector<vector<string>> brute(const string& startWord,
                             const string& targetWord,
                             const vector<string>& wordList) {
    unordered_set<string> wordSet(wordList.begin(), wordList.end());
    int INF = INT_MAX;
    int minStep = INF;
    vector<vector<string>> minList;

    function<void(string,int,unordered_set<string>&,vector<string>&)>
    dfs = [&](string word, int step,
              unordered_set<string>& visited, vector<string>& path) {
        if (word == targetWord) {
            if (step < minStep) {
                minStep = step;
                minList.clear();
            }
            if (step == minStep) {
                minList.push_back(path);
            }
            return;
        }
        for (auto &next : wordList) {
            if (!visited.count(next) && diff_by_one_2(word, next)) {
                visited.insert(next);
                path.push_back(next);
                dfs(next, step + 1, visited, path);
                path.pop_back();
                visited.erase(next);
            }
        }
    };

    unordered_set<string> visited;
    visited.insert(startWord);
    vector<string> path = {startWord};
    dfs(startWord, 1, visited, path);
    return minList;
}

// Optimal: BFS + parents graph + backtracking
vector<vector<string>> optimal(const string& startWord,
                               const string& targetWord,
                               const vector<string>& wordList) {
    unordered_set<string> wordSet(wordList.begin(), wordList.end());
    if (!wordSet.count(targetWord)) return {};

    queue<string> q;
    q.push(startWord);
    unordered_map<string, vector<string>> parents;
    unordered_set<string> visited;
    visited.insert(startWord);
    bool found = false;

    while (!q.empty() && !found) {
        int sz = q.size();
        unordered_set<string> next_level;
        for (int _ = 0; _ < sz; _++) {
            string word = q.front(); q.pop();
            for (int i = 0; i < (int)word.size(); i++) {
                string temp = word;
                for (char ch = 'a'; ch <= 'z'; ch++) {
                    temp[i] = ch;
                    if (wordSet.count(temp) && !visited.count(temp)) {
                        parents[temp].push_back(word);
                        if (temp == targetWord) found = true;
                        next_level.insert(temp);
                    }
                }
            }
        }
        for (auto &w : next_level) {
            visited.insert(w);
            q.push(w);
        }
    }

    vector<vector<string>> res;
    if (!found) return res;

    function<void(const string&, vector<string>&)> backtrack =
        [&](const string& word, vector<string>& path) {
            if (word == startWord) {
                vector<string> rev = path;
                reverse(rev.begin(), rev.end());
                res.push_back(rev);
                return;
            }
            for (auto &p : parents[word]) {
                path.push_back(p);
                backtrack(p, path);
                path.pop_back();
            }
        };

    vector<string> path = {targetWord};
    backtrack(targetWord, path);
    return res;
}

int main() {
    string startWord = "hit";
    string targetWord = "cog";
    vector<string> wordList = {"hot","dot","dog","lot","log","cog"};

    cout << "Brute:\n";
    auto r1 = brute(startWord, targetWord, wordList);
    for (auto &seq : r1) {
        for (size_t i = 0; i < seq.size(); i++) {
            cout << seq[i];
            if (i + 1 < seq.size()) cout << "->";
        }
        cout << "\n";
    }

    cout << "\nOptimal:\n";
    auto r2 = optimal(startWord, targetWord, wordList);
    for (auto &seq : r2) {
        for (size_t i = 0; i < seq.size(); i++) {
            cout << seq[i];
            if (i + 1 < seq.size()) cout << "->";
        }
        cout << "\n";
    }
    return 0;
}
